common/client/multi_node.go

```
@@ -104,101 +68,116 @@ type multiNode[
+) MultiNode[CHAIN_ID, BLOCK_HASH, HEAD, RPC_CLIENT] {
+	// TODO: does node selector only need primary nodes, or all nodes?
+	nodeSelector := newNodeSelector(selectionMode, primaryNodes)
 	// Prometheus' default interval is 15s, set this to under 7.5s to avoid
 	// aliasing (see: https://en.wikipedia.org/wiki/Nyquist_frequency)
 	const reportInterval = 6500 * time.Millisecond
-	if sendTxSoftTimeout == 0 {
-		sendTxSoftTimeout = QueryTimeout / 2
-	}
-	c := &multiNode[CHAIN_ID, SEQ, ADDR, BLOCK_HASH, TX, TX_HASH, EVENT, EVENT_OPS, TX_RECEIPT, FEE, HEAD, RPC_CLIENT, BATCH_ELEM]{
-		nodes:               nodes,
-		sendonlys:           sendonlys,
-		chainID:             chainID,
-		chainType:           chainType,
-		lggr:                logger.Sugared(lggr).Named("MultiNode").With("chainID", chainID.String()),
-		selectionMode:       selectionMode,
-		noNewHeadsThreshold: noNewHeadsThreshold,
-		nodeSelector:        nodeSelector,
-		chStop:              make(services.StopChan),
-		leaseDuration:       leaseDuration,
-		chainFamily:         chainFamily,
-		classifySendTxError: classifySendTxError,
-		reportInterval:      reportInterval,
-		sendTxSoftTimeout:   sendTxSoftTimeout,
+	c := &multiNode[CHAIN_ID, BLOCK_HASH, HEAD, RPC_CLIENT]{
+		primaryNodes:   primaryNodes,
+		sendOnlyNodes:  sendOnlyNodes,
+		chainID:        chainID,
+		lggr:           logger.Sugared(lggr).Named("MultiNode").With("chainID", chainID.String()),
+		selectionMode:  selectionMode,
+		nodeSelector:   nodeSelector,
+		chStop:         make(services.StopChan),
+		leaseDuration:  leaseDuration,
+		chainFamily:    chainFamily,
+		reportInterval: reportInterval,
 	}
 
 	c.lggr.Debugf("The MultiNode is configured to use NodeSelectionMode: %s", selectionMode)
 
 	return c
 }
 
+func (c *multiNode[CHAIN_ID, BLOCK_HASH, HEAD, RPC_CLIENT]) DoAll(ctx context.Context, do func(ctx context.Context, rpc RPC_CLIENT, isSendOnly bool) bool) error {
```

we should also return an error if we made 0 `do` calls for primary nodes